src/notebook/model.ts

```diff
@@ -642,209 +475,97 @@ class NotebookModel implements INotebookModel {
    * set of metadata on the notebook.
    */
   getMetadata(name: string): IMetadataCursor {
-    let invalid = ['kernelspec', 'languageInfo', 'origNbformat'];
-    if (invalid.indexOf(name) !== -1) {
-      let key = invalid[invalid.indexOf(name)];
-      throw Error(`Use model attribute for ${key} directly`);
-    }
-    return new MetadataCursor(
+    let cursor = new MetadataCursor(
       name,
-      this._metadata,
-      this._cursorCallback.bind(this)
+      () => {
+        return this._metadata[name];
+      },
+      (value: string) => {
+        this.setCursorData(name, value);
+      }
     );
+    this._cursors.push(cursor);
+    return cursor;
   }
 
   /**
    * List the metadata namespace keys for the notebook.
-   *
-   * #### Notes
-   * Metadata associated with the nbformat are not included.
    */
   listMetadata(): string[] {
     return Object.keys(this._metadata);
   }
 
   /**
-   * Handle changes to cell editors.
-   */
-  protected onEditorChanged(editor: IEditorModel, args: IChangedArgs<any>): void {
-    if (args.name === 'text') {
-      this.dirty = true;
-    }
-  }
-
-  /**
-   * Handle an edge request from a text editor.
-   */
-  protected onEdgeRequested(sender: IEditorModel, args: EdgeLocation): void {
-    switch (args) {
-    case 'top':
-      this.activeCellIndex--;
-      break;
-    case 'bottom':
-      this.activeCellIndex++;
-      break;
-    }
-  }
-
-  /**
    * Handle a change in the cells list.
    */
-  protected onCellsChanged(list: ObservableList<ICellModel>, change: IListChangedArgs<ICellModel>): void {
+  protected onCellsChanged(list: IObservableList<ICellModel>, change: IListChangedArgs<ICellModel>): void {
```

Comment:
I think we need to carefully audit the life cycle of the `this.onCellChanged` handler. For example, when a cell is removed, replaced, or set, is the signal handler disconnected?